feat: use pnpm to manage pm2

ojeytonwilliams opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

~@raisedadead this doesn't occur in development because `babel-node` adds the NODE_PATH (by some magic).~

~I'm not very happy with this workaround, but it's the best I've come up with so far.~

I've reverted that change in favour of using `pnpm` to manage `pm2` (explained below and repeated here for visibility).  With this change `pnpm` will update the `NODE_PATH` variable when it calls `pm2`, allowing `loopback-boot` to find the 'missing' package.

In addition, this approach means `pm2` runs using the node version that `pnpm` provides. So, we can set the Node version once (in `.npmrc`) and it can be used across all workspaces and all environments. At least in principle: for now it just ensures that the api process is managed the same way in dev, CI and production.

### Tasks

- [x] Install pm2 locally for api
- [x] Set Node version in .npmrc
- [x] Update api docs
- [ ] Update api pipelines
- [ ] Update client pipelines

My original plan was to use pnpm to manage pm2 for the client, but it can wait. 

We'd gain some reliability if we could guarantee that the production setup was the same as CI, but that's considerably more involved than just using pnpm in the VM. Worth it? Arguably, but it's not strictly necessary for deployment.

<!-- Feel free to add any additional description of changes below this line -->

ojeytonwilliams committed:
fix: show loopback where strong-error-handler is

ojeytonwilliams commented:
From what I can see from poking around in the docs, the problem is the following. While booting Loopback checks to see if all the modules it needs can be resolved. It can't find one it needs, so it throws an error.

This happens because it doesn't look in all the places node _actually_ looks, and so doesn't realise that `require` would work.

ShaunSHamilton commented:
I cannot find literature on this: I would expect explicitly setting `NODE_PATH` to cause Node to ignore other paths.

That is, I would expect this to cause Node to only resolve (find) level 2 dependencies (our dependencies' dependencies).

ojeytonwilliams commented:
I believe it adds an additional place to check, but I'll confirm on Monday.

Also, it occurs to me we should be using pm2 to manage our server during CI. If we had been doing, I'd have discovered the issue in the first PR.

naomi-lgbt commented:
> If the NODE_PATH environment variable is set to a colon-delimited list of absolute paths, then Node.js will search those paths for modules if they are not found elsewhere.

It should be an extra place to look yes

ojeytonwilliams commented:
@ShaunSHamilton it's a missing dependency, but it only shows up if you invoke `node /lib/production-start.js` directly. I'm not sure why, but `(p)npm start` works just fine.

`pm2` won't run `npm start` in cluster mode, though, so it would waste all but one core in the VM.

ojeytonwilliams committed:
Revert "fix: show loopback where strong-error-handler is"

This reverts commit aac067bd7d81cce69b3151510264d7bb8231510e.

ojeytonwilliams committed:
fix(api): use pnpm to manage pm2

This means that pm2 will get the correct NODE_PATH environment variable.
Also, if we use pnpm to manage node, this ensure that pm2 will use that
node version.

ojeytonwilliams committed:
docs(devops): update api install and update docs

ojeytonwilliams committed:
feat: set node version via npmrc

This means that all scripts (pnpm run ...) will use this node version.
The only way to get other node versions is to invoke `node` directly.

ojeytonwilliams committed:
fix(docs): update api docs for pnpm installation

ojeytonwilliams commented:
@raisedadead I'm slowly getting all the ideas straight, sorry for the delay.

I have a couple of questions for you:

1. Are you okay with running _all_ the `pm2` commands inside `api-server`? If you are, I'll update the docs accordingly. If you'd rather run everything from root, I'll install `pm2` there instead.
2. I'd like to fix the node version via `.npmrc`, but that doesn't exist on the client vms. Are you okay with using lts node inside those vms?

raisedadead commented:
> 1. Are you okay with running _all_ the `pm2` commands inside `api-server`? If you are, I'll update the docs accordingly. If you'd rather run everything from root, I'll install `pm2` there instead.

I don't care as long as I know how it works. So, Yes – it is fine wherever it is easier.

> 2. I'd like to fix the node version via `.npmrc`, but that doesn't exist on the client vms. Are you okay with using lts node inside those vms?

That should be OK.

ojeytonwilliams committed:
feat: use pm2 from root

ojeytonwilliams commented on api-server/package.json:
```diff
@@ -22,7 +22,8 @@
     "build": "babel src --out-dir lib --ignore '/**/*.test.js' --copy-files --no-copy-ignored",
     "develop": "node src/development-start.js",
-    "start": "cross-env DEBUG=fcc* node lib/production-start.js"
+    "start": "cross-env DEBUG=fcc* node lib/production-start.js",
+    "start-pm2": "pm2 start ecosystem.config.js"
```
```suggestion
    "start": "cross-env DEBUG=fcc* node lib/production-start.js"
```

ojeytonwilliams committed:
refactor: drop unused script

ShaunSHamilton commented on docs/devops.md:
````diff
@@ -534,13 +535,13 @@ pnpm run create:config && pnpm run build:curriculum && pnpm run build:server
 
 ```console
-cd api-server && pm2 reload ecosystem.config.js && cd .. && pm2 logs
+pnpm pm2 reload ecosystem.config.js && pnpm pm2 logs
````
?

```suggestion
pnpm pm2 reload api-server/ecosystem.config.js && pnpm pm2 logs
```

ShaunSHamilton reviewed (dismissed):
EELAR

ojeytonwilliams committed:
fix: use correct path to ecosystem.config.js

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ShaunSHamilton commented on docs/devops.md:
````diff
@@ -450,56 +450,57 @@ sudo apt install build-essential
 
    ```console
-   pnpm run prebuild && pnpm run build:curriculum && pnpm run build:server
+   pnpm prebuild && pnpm build:curriculum && pnpm build:server
````
Why does this not have the `run`, but the one below does?

ojeytonwilliams replied:
I forgot to update the other one, I'll do that in a followup PR.

ShaunSHamilton reviewed (approved):
LGTM 👍

naomi-lgbt reviewed (approved):
Should be good!
